src/transformers/utils/quantizers.py

```diff
@@ -0,0 +1,741 @@
+                quantization_status = "prequantized"
+
+        elif self.quantization_method is not None:
+            quantization_status = "fresh"
+
+        else:
+            # Normal loading without quantization. No quant arguments present in args or model config.
+            return
+
+        # set quantizer and config classes ========================================================================
+        if self.quantization_method == QuantizationMethod.BITS_AND_BYTES:
+            config_class = BitsAndBytesConfig
+            if dict(self.quantization_config).get("load_in_4bit", False):
+                quantizer_class = Bnb4BitHFQuantizer
+            elif dict(self.quantization_config).get("load_in_8bit", False):
+                quantizer_class = Bnb8BitHFQuantizer
+            else:
+                raise ValueError("BnB config should specify either `load_in_4bit` or `load_in_8bit`")
+
+        elif self.quantization_method == QuantizationMethod.GPTQ:
+            quantizer_class = GPTQHFQuantizer
+            config_class = GPTQConfig
+
+        else:
+            # unknown quantization_method, which is not None:
+            raise NotImplementedError(
+                f"Unsupported quantization method detected: {self.quantization_method}. Check for updates."
+            )
+
+        # finishing touches ========================================================
+        if not isinstance(self.quantization_config, (dict, QuantizationConfigMixin)):
+            raise ValueError(
+                f"Invalid type for `quantization_config`: {type(self.quantization_config)}. Should be a `dict` or a"
+                " `QuantizationConfigMixin` subclass instance."
+            )
+
+        self.quantization_config = config_class.from_dict(
+            config_dict=dict(self.quantization_config)
+        )  # dict to class instance
+
+        config.quantization_config = self.quantization_config
+        quantizer = quantizer_class(
+            quantization_config=self.quantization_config, quantization_status=quantization_status
+        )
+        return quantizer
+
+
+def get_module_from_name(module, tensor_name):
+    if "." in tensor_name:
+        splits = tensor_name.split(".")
+        for split in splits[:-1]:
+            new_module = getattr(module, split)
+            if new_module is None:
+                raise ValueError(f"{module} has no attribute {split}.")
+            module = new_module
+        tensor_name = splits[-1]
+    return module, tensor_name
+
+
+class HFQuantizer(ABC):
+    requires_parameters_quantization = False
+    aux_keys_suffixes = ()
+
+    @property
+    @abstractmethod
+    def method(self):
+        """placeholder to method from QuantizationMethod"""
+        ...
+
+    def __init__(self, quantization_config, quantization_status, **kwargs):
+        self.quantization_status = quantization_status
+        self.modules_to_not_convert = []
+
+    def validate_environment(self, *args, **kwargs):
```

Comment:
```suggestion
    @abstractmethod
    def validate_environment(self, *args, **kwargs):
```

Reply:
Could we keep it as is?  Some `HFQuantizer` subclasses may choose not to care about the env and not redefine this method.

Reply:
Hmm we could always let users define a "dummy" `validate_environment` right?

Reply:
agreed. I see benefits in the need for the developers to double-check their environment requirements.

Reply:
Makes sense!